js/web/lib/wasm/wasm-training-core-impl.ts

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import {InferenceSession} from 'onnxruntime-common';
+
+import {SerializableModeldata, SerializableSessionMetadata} from './proxy-messages';
+import {setSessionOptions} from './session-options';
+import {getInstance} from './wasm-factory';
+import {checkLastError} from './wasm-utils';
+
+const NO_TRAIN_FUNCS_MSG =
+    'Built without training API\'s enabled. Use the onnxruntime-web/training import for training ' +
+    'functionality, and make sure that all the correct artifacts are built & moved to the correct folder if ' +
+    'using a custom build. Check https://onnxruntime.ai/docs/build/web.html for more information.';
+
+export const createCheckpointHandle = (checkpointData: SerializableModeldata): number => {
+  const wasm = getInstance();
+
+  const [checkpointDataOffset, checkpointDataLength] = checkpointData;
+  let checkpointHandle = 0;
+
+  try {
+    if (wasm._OrtTrainingLoadCheckpoint) {
+      checkpointHandle = wasm._OrtTrainingLoadCheckpoint(checkpointDataOffset, checkpointDataLength);
+    } else {
+      throw new Error(NO_TRAIN_FUNCS_MSG);
+    }
+
+    if (checkpointHandle === 0) {
+      checkLastError('Error occurred when trying to create a CheckpointState.');
+    }
+    return checkpointHandle;
+  } catch (e) {
+    if (wasm._OrtTrainingReleaseCheckpoint && checkpointHandle !== 0) {
+      wasm._OrtTrainingReleaseCheckpoint(checkpointHandle);
+    }
+    throw e;
+  } finally {
+    // free buffer from wasm heap
+    wasm._OrtFree(checkpointData[0]);
+  }
+};
+
+const getTrainingModelInputOutputCount = (trainingSessionId: number): [number, number] => {
+  const wasm = getInstance();
+  const stack = wasm.stackSave();
+  try {
+    const dataOffset = wasm.stackAlloc(8);
+    if (wasm._OrtTrainingGetInputOutputCount) {
+      const errorCode = wasm._OrtTrainingGetInputOutputCount(trainingSessionId, dataOffset, dataOffset + 4, false);
+      if (errorCode !== 0) {
+        checkLastError('Can\'t get session input/output count.');
+      }
+      return [wasm.HEAP32[dataOffset / 4], wasm.HEAP32[dataOffset / 4 + 1]];
+    } else {
+      throw new Error(NO_TRAIN_FUNCS_MSG);
+    }
+  } finally {
+    wasm.stackRestore(stack);
+  }
+};
+
+const getTrainingNamesLoop = (trainingSessionId: number, count: number, isInput: boolean): [string[], number[]] => {
```

Comment:
nit: getTrainingModelInputOutputNames

Reply:
you can actually make this generic and add a boolean to specify whether you want training model or eval model
```suggestion
const getModelInputOutputNamesLoop = (trainingSessionId: number, count: number, isInput: boolean, IsEvalModel:boolean): [string[], number[]] => {
```